bundle/filefs.go

```diff
@@ -0,0 +1,73 @@
+// +build go1.16
+
+package bundle
+
+import (
+	"fmt"
+	"io"
+	"io/fs"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+type dirLoaderFS struct {
+	filesystem fs.FS
+	files      []string
+	idx        int
+}
+
+// Interface guard for dirLoaderFS to validate that it implements the DirectoryLoader interface
+var _ DirectoryLoader = (*dirLoader)(nil)
+
+// NewFSLoader returns a basic DirectoryLoader implementation
+// that will load files from a fs.FS interface
+func NewFSLoader(filesystem fs.FS) DirectoryLoader {
+	d := dirLoaderFS{
+		filesystem: filesystem,
+	}
+
+	return &d
+}
+
+// NextFile iterates to the next file in the directory tree
+// and returns a file Descriptor for the file.
+func (d *dirLoaderFS) NextFile() (*Descriptor, error) {
+	// build a list of all files we will iterate over and read, but only one time
+	if d.files == nil {
+		d.files = []string{}
+		err := fs.WalkDir(d.filesystem, ".", func(path string, dirEntry fs.DirEntry, err error) error {
+			if err != nil {
+				return err
+			}
+
+			if dirEntry != nil && dirEntry.Type().IsRegular() {
+				pathWithSlash := fmt.Sprintf("/%s", path)
+				d.files = append(d.files, filepath.ToSlash(pathWithSlash))
+			}
+
+			return nil
+		})
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to list files")
+		}
+	}
+
+	// If done reading files then just return io.EOF
+	// errors for each NextFile() call
+	if d.idx >= len(d.files) {
+		return nil, io.EOF
+	}
+
+	fileName := d.files[d.idx]
+	fileNameWithoutSlash := strings.TrimPrefix(fileName, "/")
+	d.idx++
+	fh, err := d.filesystem.Open(fileNameWithoutSlash)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to open file %s", fileNameWithoutSlash)
```

Comment:
[nit] From go 1.13 on, we can do this to the same effect:
```suggestion
		return nil, fmt.Errorf("failed to open file %s: %w", fileNameWithoutSlash, err)
```